core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/internal/signer/BaseAwsCrtV4aHttpSigner.java

```diff
@@ -0,0 +1,189 @@
+import software.amazon.awssdk.crt.auth.signing.AwsSigningResult;
+import software.amazon.awssdk.crt.http.HttpRequest;
+import software.amazon.awssdk.http.ContentStreamProvider;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.http.auth.aws.crt.AwsCrtV4aHttpSigner;
+import software.amazon.awssdk.http.auth.aws.util.CredentialUtils;
+import software.amazon.awssdk.http.auth.spi.AsyncSignRequest;
+import software.amazon.awssdk.http.auth.spi.AsyncSignedRequest;
+import software.amazon.awssdk.http.auth.spi.SignRequest;
+import software.amazon.awssdk.http.auth.spi.SyncSignRequest;
+import software.amazon.awssdk.http.auth.spi.SyncSignedRequest;
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+
+/**
+ * An internal extension of {@link AwsCrtV4aHttpSigner} that enables composable implementations of aws-signers that use
+ * a set of properties, which may extend {@link AwsCrtV4aProperties}, in order to sign requests.
+ * <p>
+ * The process for signing requests to AWS services is documented
+ * <a href="https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_aws-signing.html">here</a>.
+ */
+@SdkInternalApi
+public interface BaseAwsCrtV4aHttpSigner<T extends AwsCrtV4aProperties> extends AwsCrtV4aHttpSigner {
+
+    /**
+     * Get the base implementation of a {@link BaseAwsCrtV4aHttpSigner} that uses {@link AwsCrtV4aProperties}.
+     */
+    static BaseAwsCrtV4aHttpSigner<AwsCrtV4aProperties> create() {
+        return new BaseAwsCrtV4aHttpSignerImpl();
+    }
+
+    @Override
+    default SyncSignedRequest sign(SyncSignRequest<? extends AwsCredentialsIdentity> request) {
+        T properties = getProperties(request);
+
+        // anonymous credentials, don't sign
+        if (CredentialUtils.isAnonymous(request.identity())) {
+            return SyncSignedRequest.builder()
+                .request(request.request())
+                .payload(request.payload().orElse(null))
+                .build();
+        }
+
+        AwsSigningConfig signingConfig = createSigningConfig(properties);
+
+        SdkHttpRequest sanitizedRequest = adaptRequest(request.request(), properties);
+
+        HttpRequest crtRequest = transformRequest(sanitizedRequest, request.payload().orElse(null), properties);
+
+        SigV4aRequestContext sigV4aRequestContext = sign(sanitizedRequest, crtRequest, signingConfig, properties);
+
+        ContentStreamProvider payload = sign(request.payload().orElse(null), sigV4aRequestContext, properties);
+
+        return SyncSignedRequest.builder()
+            .request(sigV4aRequestContext.getSignedRequest())
+            .payload(payload)
+            .build();
+    }
+
+    @Override
+    default AsyncSignedRequest signAsync(AsyncSignRequest<? extends AwsCredentialsIdentity> request)
+            throws UnsupportedOperationException {
+        // There isn't currently a concept of async for crt signers
+        throw new UnsupportedOperationException();
```

Comment:
How does it work for async streaming operations, today?

Reply:
The crt signers don't have async support today.